Fix race condition and make CUDA kthvalue deterministic

nick-kuhn opened the pull request:
The gatherKthValue kernel had a race condition where multiple threads could write to the same output location without synchronization when duplicate k-th values exist, resulting in non-deterministic output.

Changes:
- aten/src/ATen/native/cuda/Sorting.cu: Use atomicMin with shared memory to deterministically find minimum index. Add early termination and remove redundant inRange checks. (We have to cast the index to `int32_t`, but this is already assumed to fit earlier in the kernel.)
- aten/src/ATen/native/cuda/Sorting.cpp: Remove non-deterministic alert since kthvalue is now deterministic on CUDA.
- torch/__init__.py: Remove kthvalue from non-deterministic operations list and remove kthvalue example from use_deterministic_algorithms() docstring.
- test/test_torch.py: Remove test_nondeterministic_alert_kthvalue since kthvalue no longer raises alerts on CUDA.

Benefits:
- Deterministic: always returns minimum index when duplicates exist
- Potential performance improvement on large arrays with repetitions

Test Results:
- All existing PyTorch tests pass (test_kthvalue)
- Custom determinism tests confirm consistent results
- Custom CUDA vs CPU correctness validated across 50+ scenarios
- Custom performance benchmarks show improvements with no visible regressions

Addresses #165227

nick-kuhn committed:
Fix race condition and make CUDA kthvalue deterministic

The gatherKthValue kernel had a race condition where multiple threads
could write to the same output location without synchronization when
duplicate k-th values exist, resulting in non-deterministic output.

Changes:
- aten/src/ATen/native/cuda/Sorting.cu: Use atomicMin with shared memory
  to deterministically find minimum index. Add early termination and remove
  redundant inRange checks. (We have to cast the index to `int32_t`, but this is already assumed to fit earlier in the kernel.)
- aten/src/ATen/native/cuda/Sorting.cpp: Remove non-deterministic alert
  since kthvalue is now deterministic on CUDA.
- torch/__init__.py: Remove kthvalue from non-deterministic operations list
  and remove kthvalue example from use_deterministic_algorithms() docstring.
- test/test_torch.py: Remove test_nondeterministic_alert_kthvalue since
  kthvalue no longer raises alerts on CUDA.

Benefits:
- Deterministic: always returns minimum index when duplicates exist
- Faster: 10-23% performance improvement on large arrays
- Correct: eliminates undefined behavior from concurrent writes

Test Results:
- All existing PyTorch tests pass (test_kthvalue)
- Custom determinism tests confirm consistent results
- Custom CUDA vs CPU correctness validated across 50+ scenarios
- Custom erformance benchmarks show improvements with no visible regressions

nick-kuhn commented:
@pytorchbot label "module: determinism", "release notes: cuda"

ngimel commented on aten/src/ATen/native/cuda/Sorting.cu:
```diff
@@ -65,25 +65,34 @@ __global__ void gatherKthValue(
   // Find the index of the k-th highest element
-  index_t kValueIndex = 0;
-  bool foundKValue = false;
+  __shared__ int32_t minIndexFound;
```
this (and atomic below) won't correctly work when the kernel is launched with `index_t = int64_t`, for large slices that require 64-bit indexing

ngimel replied:
Thanks for checking! Can you please error out in this case then? Does instantiating with index_t = int64_t ever make sense, or will it only lead to silent wrong results?

nick-kuhn committed:
Add slice size validation for kthvalue CUDA op

Check that slice size does not exceed int32_t max, as this is assumed by the current implementation.

nick-kuhn committed:
Fix linting issues (trailing spaces)

nick-kuhn committed:
Merge branch 'pytorch:main' into feat/kthvalue-kernel-determinism
